### PR TITLE
Добавить возможность теневой параллельной стратегии инициализации

### PR DESCRIPTION
**Pull request by xakep139**
This change is [Reviewable](https://reviewable.io/reviews/2gis/nuclear-river/138)

### CODE REVIEW

**Comment by denisivan0v** [comment]
Reviewed 14 of 14 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[StateInitialization/StateInitialization.Core/Actors/BulkInsertDataObjectsActor.cs, line 36 at r1](https://reviewable.io:443/reviews/2gis/nuclear-river/138#-KYDdyNmqMioGeSwG2zS:-KYDdyNmqMioGeSwG2zT:bi0djv0) ([raw file](https://github.com/2gis/nuclear-river/blob/68104da08376265b10e4f552d3cf4022c64fa5c3/StateInitialization/StateInitialization.Core/Actors/BulkInsertDataObjectsActor.cs#L36)):*
> command.TargetTable.Table
Можеть быть передавать только префикс?

---

*[StateInitialization/StateInitialization.Core/Actors/CreateTableCopyActor.cs, line 108 at r1](https://reviewable.io:443/reviews/2gis/nuclear-river/138#-KYDlp5X-FLK52eae0Fi:-KYDlp5X-FLK52eae0Fj:bmmailk) ([raw file](https://github.com/2gis/nuclear-river/blob/68104da08376265b10e4f552d3cf4022c64fa5c3/StateInitialization/StateInitialization.Core/Actors/CreateTableCopyActor.cs#L108)):*
> $"[{sourceTableName}]"
Есть функции для генерации правильного названия таблицы (квадратные скобки). Либо у SqlConnection, либо в SMO, либо в Linq2Db

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/2gis/nuclear-river/138)*

**Comment by xakep139** [comment]
Review status: 13 of 18 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[StateInitialization/StateInitialization.Core/Actors/BulkInsertDataObjectsActor.cs, line 36 at r1](https://reviewable.io:443/reviews/2gis/nuclear-river/138#-KYDdyNmqMioGeSwG2zS:-KYDo3iZOhmfbGFVSxKE:bb7wlak) ([raw file](https://github.com/2gis/nuclear-river/blob/68104da08376265b10e4f552d3cf4022c64fa5c3/StateInitialization/StateInitialization.Core/Actors/BulkInsertDataObjectsActor.cs#L36)):*
<details><summary><i>Previously, denisivan0v (Denis Ivanov) wrote…</i></summary><blockquote>

> command.TargetTable.Table
Можеть быть передавать только префикс?
</blockquote></details>

Тогда Actor должен будет знать про стороннюю логику

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/2gis/nuclear-river/138)*

**Comment by xakep139** [comment]
Review status: 12 of 18 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[StateInitialization/StateInitialization.Core/Actors/CreateTableCopyActor.cs, line 108 at r1](https://reviewable.io:443/reviews/2gis/nuclear-river/138#-KYDlp5X-FLK52eae0Fi:-KYIHuJXCl2Dt61ADae5:b-qofxva) ([raw file](https://github.com/2gis/nuclear-river/blob/68104da08376265b10e4f552d3cf4022c64fa5c3/StateInitialization/StateInitialization.Core/Actors/CreateTableCopyActor.cs#L108)):*
<details><summary><i>Previously, denisivan0v (Denis Ivanov) wrote…</i></summary><blockquote>

> $"[{sourceTableName}]"
Есть функции для генерации правильного названия таблицы (квадратные скобки). Либо у SqlConnection, либо в SMO, либо в Linq2Db
</blockquote></details>

Исправил в fe2c8fb

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/2gis/nuclear-river/138)*

**Comment by denisivan0v** [comment]
Reviewed 6 of 6 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[StateInitialization/StateInitialization.Core/Actors/BulkInsertDataObjectsActor.cs, line 36 at r1](https://reviewable.io:443/reviews/2gis/nuclear-river/138#-KYDdyNmqMioGeSwG2zS:-KYIJezw0fNZU8Jo-1mN:b-vaubr3) ([raw file](https://github.com/2gis/nuclear-river/blob/68104da08376265b10e4f552d3cf4022c64fa5c3/StateInitialization/StateInitialization.Core/Actors/BulkInsertDataObjectsActor.cs#L36)):*
<details><summary><i>Previously, xakep139 (Nikita Balabaev) wrote…</i></summary><blockquote>

Тогда Actor должен будет знать про стороннюю логику
</blockquote></details>

Ну как стороннюю.. Моя мысль в том, что `BulkInsertDataObjectsActor` должен вставлять данные типа `TDataObject` в ту таблицу, которая соответствует этому типу в маппинге. Дополнительно, мы можем передавать префикс в команде, и в этом случае актор будет добавлять этот префикс к имени таблицы.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/2gis/nuclear-river/138)*

File: StateInitialization/StateInitialization.Core/Actors/ConstraintsManagementActor.cs [hunk 1]
```diff
@@ -30,6 +30,16 @@ public IReadOnlyCollection<IEvent> ExecuteCommands(IReadOnlyCollection<ICommand>
             {
                 var tables = GetTables();
 
+                var defaults = tables.SelectMany(t => t.Columns.Cast<Column>().Select(c => c.DefaultConstraint))
```
Comment (xakep139) [inline StateInitialization/StateInitialization.Core/Actors/ConstraintsManagementActor.cs]: Восстанавливаем значения по умолчанию

**Comment by xakep139** [comment]
Review status: 15 of 18 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[StateInitialization/StateInitialization.Core/Actors/BulkInsertDataObjectsActor.cs, line 36 at r1](https://reviewable.io:443/reviews/2gis/nuclear-river/138#-KYDdyNmqMioGeSwG2zS:-KYIRLOX4UymcjAouMZp:b-p5mlav) ([raw file](https://github.com/2gis/nuclear-river/blob/68104da08376265b10e4f552d3cf4022c64fa5c3/StateInitialization/StateInitialization.Core/Actors/BulkInsertDataObjectsActor.cs#L36)):*
<details><summary><i>Previously, denisivan0v (Denis Ivanov) wrote…</i></summary><blockquote>

Ну как стороннюю.. Моя мысль в том, что `BulkInsertDataObjectsActor` должен вставлять данные типа `TDataObject` в ту таблицу, которая соответствует этому типу в маппинге. Дополнительно, мы можем передавать префикс в команде, и в этом случае актор будет добавлять этот префикс к имени таблицы.
</blockquote></details>

Исправил в 1a38767

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/2gis/nuclear-river/138)*

**Comment by rechkalov** [comment]
Review status: 15 of 18 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[StateInitialization/StateInitialization.Core/Actors/CreateTableCopyActor.cs, line 53 at r3](https://reviewable.io:443/reviews/2gis/nuclear-river/138#-KYIiBQ5Cv8XeGUwaf-C:-KYIiBQ56t8QLnjDSnkk:b-e98pp8) ([raw file](https://github.com/2gis/nuclear-river/blob/1a3876729eefa986ccfb345677f964dc53a2bd94/StateInitialization/StateInitialization.Core/Actors/CreateTableCopyActor.cs#L53)):*
> ```Smalltalk
>             var newTable = string.IsNullOrEmpty(targetTable.Schema)
>                 ? new Table(_database, targetTable.Table)
>                 : new Table(_database, targetTable.Table, targetTable.Schema ?? "dbo");
> ```

Наверное, планировалось:
```
string.IsNullOrEmpty(targetTable.Schema)
    ? new Table(_database, targetTable.Table, "dbo")
    : new Table(_database, targetTable.Table, targetTable.Schema);
```

---

*[StateInitialization/StateInitialization.Core/Actors/CreateTableCopyActor.cs, line 99 at r3](https://reviewable.io:443/reviews/2gis/nuclear-river/138#-KYIkfXh45aobfXpsJ6K:-KYIkfXi2Ew_y9Er4P-y:bls8kg9) ([raw file](https://github.com/2gis/nuclear-river/blob/1a3876729eefa986ccfb345677f964dc53a2bd94/StateInitialization/StateInitialization.Core/Actors/CreateTableCopyActor.cs#L99)):*
> ```Smalltalk
>         }
> 
>         private Index CreateIndexCopy(Table targetTable, Index index)
> ```

Зачем в копии создавать индексы? 
Приходится потом их отключать, потом включать, потом переименовывать. Можно ли при замене таблицы - у исходной получить индексы, сериализовать методом Script() и создать сразу на новой таблице?

Можно даже пойти дальше и не отключать индексы и потом включать, а реально удалять - тогда этот шаг не будет отличаться от других стратегий репликации.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/2gis/nuclear-river/138)*

**Comment by xakep139** [comment]
Review status: 15 of 18 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[StateInitialization/StateInitialization.Core/Actors/CreateTableCopyActor.cs, line 53 at r3](https://reviewable.io:443/reviews/2gis/nuclear-river/138#-KYIiBQ5Cv8XeGUwaf-C:-KYLut37B0TMxyYLorfU:bzh1u8l) ([raw file](https://github.com/2gis/nuclear-river/blob/1a3876729eefa986ccfb345677f964dc53a2bd94/StateInitialization/StateInitialization.Core/Actors/CreateTableCopyActor.cs#L53)):*
<details><summary><i>Previously, rechkalov (a.rechkalov) wrote…</i></summary><blockquote>

Наверное, планировалось:
```
string.IsNullOrEmpty(targetTable.Schema)
    ? new Table(_database, targetTable.Table, "dbo")
    : new Table(_database, targetTable.Table, targetTable.Schema);
```
</blockquote></details>

dbo является схемой по умолчанию не всегда и не для каждого пользователя, тем более если мы будем использовать источник или назначение, отличное от MS SQL Server

---

*[StateInitialization/StateInitialization.Core/Actors/CreateTableCopyActor.cs, line 99 at r3](https://reviewable.io:443/reviews/2gis/nuclear-river/138#-KYIkfXh45aobfXpsJ6K:-KYLv5ZW9OCWrIM5f-rU:b-q8g05e) ([raw file](https://github.com/2gis/nuclear-river/blob/1a3876729eefa986ccfb345677f964dc53a2bd94/StateInitialization/StateInitialization.Core/Actors/CreateTableCopyActor.cs#L99)):*
<details><summary><i>Previously, rechkalov (a.rechkalov) wrote…</i></summary><blockquote>

Зачем в копии создавать индексы? 
Приходится потом их отключать, потом включать, потом переименовывать. Можно ли при замене таблицы - у исходной получить индексы, сериализовать методом Script() и создать сразу на новой таблице?

Можно даже пойти дальше и не отключать индексы и потом включать, а реально удалять - тогда этот шаг не будет отличаться от других стратегий репликации.
</blockquote></details>

Если не создать первичный ключ, то создание его после заполнения таблицы занимает очень длительное время, поскольку БД начинает физически перемещать строки каждой таблицы.
Можно, конечно, создать сначала только первичный ключ, а потом все остальные, но операция получается тогда какая-то размазанная.
Изначально я так и сделал: создавать индексы во время замены таблиц. Но тогда простой системы занимает непозволительно большое время (порядка 25 минут вместо 40 секунд сейчас), опять же из-за первичного ключа.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/2gis/nuclear-river/138)*

**Comment by xakep139** [comment]
Review status: 15 of 18 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[StateInitialization/StateInitialization.Core/Actors/CreateTableCopyActor.cs, line 53 at r3](https://reviewable.io:443/reviews/2gis/nuclear-river/138#-KYIiBQ5Cv8XeGUwaf-C:-KYLwdgp4O_E9MdImK_x:b-ezk2vh) ([raw file](https://github.com/2gis/nuclear-river/blob/1a3876729eefa986ccfb345677f964dc53a2bd94/StateInitialization/StateInitialization.Core/Actors/CreateTableCopyActor.cs#L53)):*
<details><summary><i>Previously, xakep139 (Nikita Balabaev) wrote…</i></summary><blockquote>

dbo является схемой по умолчанию не всегда и не для каждого пользователя, тем более если мы будем использовать источник или назначение, отличное от MS SQL Server
</blockquote></details>

Всё, увидел косяк )

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/2gis/nuclear-river/138)*

**Comment by xakep139** [comment]
Review status: 14 of 18 files reviewed at latest revision, 4 unresolved discussions.

---

*[StateInitialization/StateInitialization.Core/Actors/CreateTableCopyActor.cs, line 53 at r3](https://reviewable.io:443/reviews/2gis/nuclear-river/138#-KYIiBQ5Cv8XeGUwaf-C:-KYLxQ127K1coAw4nXOA:bw8fxz2) ([raw file](https://github.com/2gis/nuclear-river/blob/1a3876729eefa986ccfb345677f964dc53a2bd94/StateInitialization/StateInitialization.Core/Actors/CreateTableCopyActor.cs#L53)):*
<details><summary><i>Previously, xakep139 (Nikita Balabaev) wrote…</i></summary><blockquote>

Всё, увидел косяк )
</blockquote></details>

Исправил в 0c967f7

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/2gis/nuclear-river/138)*

**Comment by xakep139** [comment]
Review status: 14 of 18 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[StateInitialization/StateInitialization.Core/Actors/CreateTableCopyActor.cs, line 99 at r3](https://reviewable.io:443/reviews/2gis/nuclear-river/138#-KYIkfXh45aobfXpsJ6K:-KYLyf6mBlGsLP_nkLuk:b7c4b3k) ([raw file](https://github.com/2gis/nuclear-river/blob/1a3876729eefa986ccfb345677f964dc53a2bd94/StateInitialization/StateInitialization.Core/Actors/CreateTableCopyActor.cs#L99)):*
<details><summary><i>Previously, xakep139 (Nikita Balabaev) wrote…</i></summary><blockquote>

Если не создать первичный ключ, то создание его после заполнения таблицы занимает очень длительное время, поскольку БД начинает физически перемещать строки каждой таблицы.
Можно, конечно, создать сначала только первичный ключ, а потом все остальные, но операция получается тогда какая-то размазанная.
Изначально я так и сделал: создавать индексы во время замены таблиц. Но тогда простой системы занимает непозволительно большое время (порядка 25 минут вместо 40 секунд сейчас), опять же из-за первичного ключа.
</blockquote></details>

Вариант с удалением индексов в исходных таблицах тоже не подходит, поскольку это может помешать работе пользователей (в некоторых таблицах до 20 индексов).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/2gis/nuclear-river/138)*

**Comment by denisivan0v** [comment]
Reviewed 3 of 3 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/2gis/nuclear-river/138)*

**Pull request merged**